init/init.c

```diff
@@ -236,6 +238,42 @@ void init_network(const char *iface, int domain) {
     close(s);
 }
 
+// inject boot-time entropy after reading it from a vsock port
+void init_entropy(int port) {
+    int s = openvsock(VMADDR_CID_HOST, port);
+    if (s < 0) {
+        die("openvsock entropy");
+    }
+
+    int e = open("/dev/random", O_RDWR);
+    if (e < 0) {
+        die("open /dev/random");
+    }
+
+    struct {
+        int entropy_count;
+        int buf_size;
+        char buf[4096];
+    } buf;
+
+    for (;;) {
+        ssize_t n = read(s, buf.buf, sizeof(buf.buf));
```

Comment:
So I assume the host will then send/close which causes the `n == 0` break below and this returns?

Reply:
Right, when host closes we get n == 0. This loop would be a little more complicated in the presence of signals, but we disable signals early on (and I'm not sure where those signals would come from anyway).